pydicom/fileset.py

```diff
@@ -0,0 +1,2805 @@
+        parent = RecordNode(record)
+        node = parent  # Maybe only be a single record
+        for record in record_gen:
+            node = RecordNode(record)
+            node.parent = parent
+            parent = node
+
+        instance = FileInstance(node)
+        node.instance = instance
+        self._tree.add(node)
+
+        # Save the dataset to the stage
+        self._stage['+'][instance.SOPInstanceUID] = instance
+        self._instances.append(instance)
+        instance._apply_stage('+')
+        ds.save_as(instance.path, write_like_original=False)
+
+        return instance
+
+    def add_custom(
+        self, ds_or_path: Union[Dataset, str, PathLike], leaf: RecordNode
+    ) -> None:
+        """Stage an instance for addition to the File-set using custom records.
+
+        If the instance has been staged for removal then calling
+        :meth:`~pydicom.fileset.FileSet.add_custom` will cancel the staging.
+
+        Parameters
+        ----------
+        ds_or_path : pydicom.dataset.Dataset, str or PathLike
+            The instance to add to the File-set, either as a pydicom
+            :class:`~pydicom.dataset.Dataset` or the path to an instance
+            written in the :dcm:`DICOM File Format<part10/chapter_7.html>` as
+            :class:`str` or :class:`pathlib.Path`.
+        leaf : pydicom.fileset.RecordNode
+            The leaf node for the instance, should have its ancestors nodes set
+            correctly as well as their corresponding directory records. Should
+            have no more than 7 ancestors due to the semantics used by
+            :class:`~pydicom.fileset.FileSet` when creating the directory
+            structure.
+
+        Returns
+        -------
+        FileInstance
+            The :class:`~pydicom.fileset.FileInstance` that was added.
+
+        See Also
+        --------
+        :meth:`~pydicom.fileset.FileSet.add`
+        """
+        if isinstance(ds_or_path, (str, PathLike)):
+            ds = dcmread(ds_or_path)
+        else:
+            ds = ds_or_path
+
+        # Check the supplied nodes
+        if leaf.depth > 7:
+            raise ValueError(
+                "The 'leaf' node must not have more than 7 ancestors as "
+                "'FileSet' supports a maximum directory structure depth of 8"
+            )
+
+        key = ds.SOPInstanceUID
+        have_instance = [ii for ii in self if ii.SOPInstanceUID == key]
+
+        # If staged for removal, keep instead - check this now because
+        #   `have_instance` is False when instance staged for removal
+        if key in self._stage['-']:
+            instance = self._stage['-'][key]
+            del self._stage['-'][key]
+            self._instances.append(instance)
+            instance._apply_stage('+')
+
+            return instance
+
+        if have_instance:
+            return have_instance[0]
+
+        # Check the leaf node's record contains the required elements
+        keywords = [
```

Comment:
Change this so that they get set automatically if missing. They should all be available in the passed dataset and the referenced file ID isn't important when adding.